sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1032,6 +1044,42 @@ object ColumnPruning extends Rule[LogicalPlan] {
         p
       }
 
+    case p @ Project(_, ul: UnionLoop) =>
+      if (!ul.outputSet.subsetOf(p.references)) {
```

Comment:
How can this be true? The analyzer guarantees that all the references of a plan node come from its children nodes. So `ul.outputSet` must be a super set of `p.references`.